### ISSUE TITLE
Docs for `let_underscore_untyped` unclear

### ISSUE THREAD

**Issue by Fuuzetsu** [rust-lang/rust-clippy#10867]
### Description

Currently the [docs](https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_untyped) say that

`let _ = foo();` is bad because the type can change.

The suggested fix is a below

```rust
// Either provide a type annotation:
let _: Result<u32, ()> = foo();
// …or drop the let keyword:
_ = foo();
```

The first fix is fine: adding a type signature stops future surprises. However I don't understand the second suggestion. How does saying `_ = foo();` help with the issue this lint is flagging? If `foo()` type changes then `_ = foo();` still has the same problem as original `let _ = foo();`, right?

Can the docs be clarified on why `_ = foo();` is fine while `let _ = foo();` is not?

**Comment by blyxyas** [comment]
#10356
> Alternatively, they can drop the let keyword to truly just ignore the value no matter what.

`_ = foo()` would be the same as adding an `#[allow]` attribute, we should still update the documentation tho.

@rustbot claim

**Comment by nk9** [comment]
An implementation was attempted, but never got across the line. Probably a very small change. 

@rustbot label A-documentation "good first issue"

**Comment by blyxyas** [comment]
Not really "an implementation", more like we decided that it wasn't worth the effort

**Comment by nk9** [comment]
If it should be closed on that basis, that's cool too!

**Comment by Fuuzetsu** [comment]
It still doesn't make sense to me that `_ = foo()` gets special treatment. It feels like it should be linted the same as `let _ = foo()` and if the dev wants to allow the lint violation, they should just use `#[allow]` like for every other lint instead of using some other version of syntax that just happens to not throw a lint up but does nothing to mitigate the potential issue the lint exists for.

**Comment by samueltardieu** [comment]
`_ = expr;` cannot benefit from type ascription, as `_: type = expr;` is not legal Rust. It would make no sense to complain about a missing type.

Note that the suggestion to transform `let _ = expr;` into `_ = expr;` might change the semantics. For example:

```rust
fn main() {
    let foo = std::sync::Mutex::new(());
    let _ = &foo.lock().unwrap();
    println!("{:?}", *foo.lock().unwrap());
}
```

will lock forever due to lifetime extension on `let _`, while

```rust
fn main() {
    let foo = std::sync::Mutex::new(());
    _ = &foo.lock().unwrap();
    println!("{:?}", *foo.lock().unwrap());
}
```

will execute just fine as assignment to `_` doesn't benefit from lifetime extension and the `MutexGuard` is dropped after the `_ = …` statement.

But again, this is probably not an adverse effect here. However, proposing to change `_ = expr;` into `let _: type = expr;` would possibility cause adverse effect (such as blocking, in reversing the previous example) because of lifetime extension rules applying to `let _`.
